File: wf-api-integration/references/trade-order/notify-trade-order/java/controller/NotifyTradeOrderController.java
/**
 * Alipay.com Inc.
 * Copyright (c) 2004-2026 All Rights Reserved.
 */
package {basePackage}.wf.controller;

import com.alibaba.fastjson.JSON;
import {basePackage}.wf.model.request.NotifyTradeOrderRequest;
import {basePackage}.wf.model.response.NotifyTradeOrderResponse;
import {basePackage}.wf.model.response.Result;
import {basePackage}.wf.signer.WfSigner;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RestController;

/**
 * WorldFirst notifyTradeOrder 回调处理器（仅 PAY_INTO_CHINA 场景）。
 *
 * <p>WF 在交易订单处理完成后，主动 POST 到集成商提供的 notifyUrl。
 *
 * <p>处理流程：
 * <ol>
 *   <li>验签：从请求头提取 Signature，调用 WfSigner.verifySignature() 验证</li>
 *   <li>幂等判断：以 requestId 去重，已处理直接返回 SUCCESS</li>
 *   <li>异步处理：业务逻辑异步执行，避免超时触发 WF 重试</li>
 *   <li>签名响应：对响应体签名后返回</li>
 * </ol>
 *
 * <p>WF 重试策略：未收到有效响应时重试 7 次，间隔 2min→10min→10min→1h→2h→6h→15h。
 *
 * @author Qoder
 * @version NotifyTradeOrderController.java, v 0.1 2026-04-03
 */
@RestController
public class NotifyTradeOrderController {

    private static final Logger LOGGER = LoggerFactory.getLogger(NotifyTradeOrderController.class);

    private static final String NOTIFY_PATH = "/amsin/api/v1/business/account/notifyTradeOrder";
    private static final String HEADER_SIGNATURE = "Signature";
    private static final String HEADER_CLIENT_ID = "Client-Id";
    private static final String HEADER_REQUEST_TIME = "Request-Time";

    private final WfSigner signer;

    public NotifyTradeOrderController(WfSigner signer) {
        this.signer = signer;
    }

    /**
     * 处理 WF notifyTradeOrder 回调。
     *
     * @param headers HTTP 请求头
     * @param requestBody 原始请求体
     * @return 签名后的响应
     */
    @PostMapping("${wf.notify.trade-order.path:/notify/trade-order}")
    public ResponseEntity<String> handleNotification(
            @RequestHeader HttpHeaders headers,
            @RequestBody String requestBody) {

        LOGGER.info("NotifyTradeOrderController received notification");

        // Step 1: 验签
        String signature = extractSignatureValue(headers.getFirst(HEADER_SIGNATURE));
        String clientId = headers.getFirst(HEADER_CLIENT_ID);
        String requestTime = headers.getFirst(HEADER_REQUEST_TIME);

        if (signature == null || clientId == null || requestTime == null) {
            LOGGER.warn("NotifyTradeOrderController missing required headers");
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body("Missing required headers");
        }

        String signContent = "POST " + NOTIFY_PATH + "\n" + clientId + "." + requestTime + "." + requestBody;
        boolean verified = signer.verifySignature(NOTIFY_PATH, clientId, requestTime, requestBody);
        if (!verified) {
            LOGGER.warn("NotifyTradeOrderController signature verification failed");
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body("Signature verification failed");
        }

        // Step 2: 解析请求
        NotifyTradeOrderRequest request;
        try {
            request = JSON.parseObject(requestBody, NotifyTradeOrderRequest.class);
        } catch (Exception e) {
            LOGGER.error("NotifyTradeOrderController failed to parse request", e);
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body("Invalid request body");
        }

        String requestId = request.getRequestId();
        LOGGER.info("NotifyTradeOrderController processing requestId={}", requestId);

        // Step 3: 幂等判断 — 实际项目中需要持久化存储已处理的 requestId
        // if (alreadyProcessed(requestId)) {
        //     return buildSignedResponse(buildSuccessResponse());
        // }

        // Step 4: 异步处理业务逻辑 — 避免阻塞导致 WF 超时重试
        // asyncProcessTradeOrderResults(request.getTradeOrderResults());

        // Step 5: 构建成功响应
        NotifyTradeOrderResponse response = new NotifyTradeOrderResponse();
        Result result = new Result();
        result.setResultCode("SUCCESS");
        result.setResultStatus("S");
        result.setResultMessage("success");
        response.setResult(result);

        // Step 6: 签名响应
        String responseBody = JSON.toJSONString(response);
        String responseSignature = signer.generateSignature(NOTIFY_PATH, clientId, responseBody);

        HttpHeaders responseHeaders = new HttpHeaders();
        responseHeaders.set(HEADER_SIGNATURE,
            "algorithm=RSA256, keyVersion=2, signature=" + responseSignature);

        return new ResponseEntity<>(responseBody, responseHeaders, HttpStatus.OK);
    }

    /**
     * 从 Signature 头中提取签名值。
     * 格式：algorithm=RSA256, keyVersion=2, signature=xxx
     */
    private String extractSignatureValue(String signatureHeader) {
        if (signatureHeader == null || signatureHeader.isEmpty()) {
            return null;
        }
        String prefix = "signature=";
        int idx = signatureHeader.indexOf(prefix);
        if (idx < 0) {
            return null;
        }
        return signatureHeader.substring(idx + prefix.length()).trim();
    }
}

File: wf-api-integration/references/trade-order/notify-trade-order/golang/controller/notify_trade_order_controller.go
package controller

import (
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"

	"{moduleName}/wf/model/request"
	"{moduleName}/wf/model/response"
	"{moduleName}/wf/signer"
)

// NotifyTradeOrderController handles notifyTradeOrder callback notifications from WF.
// This is only applicable to the PAY_INTO_CHINA scene.
//
// After trade order processing is complete, WF sends a notifyTradeOrder
// callback via HTTP POST to the integrator's notifyUrl. This controller is responsible for:
//  1. Verifying the request signature
//  2. Parsing the request body
//  3. Idempotency check based on requestId
//  4. Processing trade order results (preferably async)
//  5. Signing and returning the response
//
// If no successful response is returned, WF will retry up to 7 times with intervals:
// 2min, 10min, 10min, 1h, 2h, 6h, 15h
type NotifyTradeOrderController struct {
	Signer signer.Signer

	// OnTradeOrderResult is called when a trade order result notification is received.
	// Integrators should implement this callback to handle business logic
	// (e.g., update local order status, process settlement results).
	OnTradeOrderResult func(req *request.NotifyTradeOrderRequest)
}

const (
	tradeOrderNotifyPath = "/amsin/api/v1/business/account/notifyTradeOrder"
)

// NewNotifyTradeOrderController creates a new NotifyTradeOrderController with default no-op callbacks.
func NewNotifyTradeOrderController(s signer.Signer) *NotifyTradeOrderController {
	return &NotifyTradeOrderController{
		Signer: s,
		OnTradeOrderResult: func(req *request.NotifyTradeOrderRequest) {
			log.Printf("[NotifyTradeOrderController] trade order results received: requestId=%s, resultCount=%d",
				req.RequestID,
				len(req.TradeOrderResults))
			for i, r := range req.TradeOrderResults {
				log.Printf("[NotifyTradeOrderController] result[%d]: referenceOrderNo=%s, orderStatus=%s, orderType=%s, statusMessage=%s",
					i, r.ReferenceOrderNo, r.OrderStatus, r.OrderType, r.StatusMessage)
			}
		},
	}
}

// ServeHTTP implements http.Handler, making NotifyTradeOrderController a standard HTTP handler.
//
// Usage with net/http:
//
//	ctrl := controller.NewNotifyTradeOrderController(wfSigner)
//	http.Handle("/notify/trade-order", ctrl)
//
// Usage with gin:
//
//	router.POST("/notify/trade-order", gin.WrapH(ctrl))
func (c *NotifyTradeOrderController) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	log.Println("[NotifyTradeOrderController] received notification")

	// Step 1: Read request body
	body, err := io.ReadAll(r.Body)
	if err != nil {
		log.Printf("[NotifyTradeOrderController] failed to read request body: %v", err)
		http.Error(w, "Failed to read request body", http.StatusBadRequest)
		return
	}
	defer r.Body.Close()

	// Step 2: Extract required headers
	signatureHeader := r.Header.Get("Signature")
	clientID := r.Header.Get("Client-Id")
	requestTime := r.Header.Get("Request-Time")

	if signatureHeader == "" || clientID == "" || requestTime == "" {
		log.Println("[NotifyTradeOrderController] missing required headers")
		http.Error(w, "Missing required headers", http.StatusBadRequest)
		return
	}

	// Step 3: Verify signature
	if err := c.Signer.VerifySignatureWithPath(tradeOrderNotifyPath, clientID, requestTime, string(body), signatureHeader); err != nil {
		log.Printf("[NotifyTradeOrderController] signature verification failed: %v", err)
		http.Error(w, "Signature verification failed", http.StatusBadRequest)
		return
	}

	// Step 4: Parse request body
	var req request.NotifyTradeOrderRequest
	if err := json.Unmarshal(body, &req); err != nil {
		log.Printf("[NotifyTradeOrderController] failed to parse request body: %v", err)
		http.Error(w, "Invalid request body", http.StatusBadRequest)
		return
	}

	if req.RequestID == "" {
		log.Println("[NotifyTradeOrderController] requestId is missing")
		http.Error(w, "requestId is missing", http.StatusBadRequest)
		return
	}

	log.Printf("[NotifyTradeOrderController] processing requestId=%s", req.RequestID)

	// Step 5: Idempotency check (integrators should implement their own deduplication logic)
	// TODO: Check if requestId has been processed before
	// if isDuplicate(req.RequestID) {
	//     log.Printf("[NotifyTradeOrderController] duplicate requestId=%s, skip processing", req.RequestID)
	//     // still return success
	// }

	// Step 6: Process business logic — async processing recommended to avoid WF timeout retry
	if c.OnTradeOrderResult != nil {
		c.OnTradeOrderResult(&req)
	}

	// Step 7: Build success response
	resp := response.NewSuccessResponse()
	respBody, _ := json.Marshal(resp)

	// Step 8: Sign response
	respSignature, err := c.Signer.GenerateSignatureWithPath(tradeOrderNotifyPath, clientID, c.Signer.GetRequestTime(), string(respBody))
	if err != nil {
		log.Printf("[NotifyTradeOrderController] failed to sign response: %v", err)
		// Still return success to avoid WF retry
	}

	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
	w.Header().Set("Signature", fmt.Sprintf("algorithm=RSA256, keyVersion=2, signature=%s", respSignature))
	w.WriteHeader(http.StatusOK)
	w.Write(respBody)

	log.Printf("[NotifyTradeOrderController] processed successfully, requestId=%s", req.RequestID)
}
